pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
+	var desiredKeys []BpfLpmIpKeySt
+	for desiredKey := range ebpfKeyToRules {
+		desiredKeys = append(desiredKeys, desiredKey)
+	}
+
+	// Get stale keys (= keys inside the eBPF map but not inside the currently desired ruleset). Those keys
+	// will be dropped in the next step.
+	staleKeys, err := infc.getStaleKeys(desiredKeys)
 	if err != nil {
-		return fmt.Errorf("Cannot get map info: %v on if %d", err, ifId)
+		return err
+	}
+
+	// Purge all stale keys from the eBPF map.
+	if err := infc.purgeKeys(staleKeys); err != nil {
+		klog.Infof("Purge keys operation encountered issues, err: %q", err)
 	}
-	klog.Infof("Ingress node firewall map Info: %+v with FD %s", info, objs.BpfMaps.IngressNodeFirewallTableMap.String())
 
-	if err := infc.makeIngressFwRulesMap(ingFireWallConfig, isDelete, ifId); err != nil {
-		return fmt.Errorf("Failed to create map firewall rules: %v on if %d", err, ifId)
+	// Add/update all keys. This is an idempotent action and non-existing keys are added whereas existing keys
+	// are updated.
+	if err := infc.addOrUpdateRules(ebpfKeyToRules); err != nil {
+		return err
 	}
 
+	// Generate ingress node few events.
 	infc.ingressNodeFwEvents()
 
 	return nil
 }
 
+// addOrUpdateRules is a small method containing this limited set of functionality to facilitate unit testing of
+// this part of code.
+func (infc *IngNodeFwController) addOrUpdateRules(ebpfKeyToRules map[BpfLpmIpKeySt]BpfRulesValSt) error {
+	for ebpfKey, ebpfRules := range ebpfKeyToRules {
+		log.Printf("Adding or updating ingress firewall rules for key %v", ebpfKey)
+		if err := infc.objs.BpfMaps.IngressNodeFirewallTableMap.Update(ebpfKey, ebpfRules, ebpf.UpdateAny); err != nil {
+			return fmt.Errorf("Failed Adding/Updating ingress firewall rules: %v", err)
+		}
+	}
+	return nil
+}
+
+// GetStatisticsMap returns the statistics map of the object.
 func (infc *IngNodeFwController) GetStatisticsMap() *ebpf.Map {
 	return infc.objs.IngressNodeFirewallStatisticsMap
 }
 
-// makeIngressFwRulesMap convert IngressNodeFirewallRules into eBPF format which matched what the
-// kerenl hook will be using.
-func (infc *IngNodeFwController) makeIngressFwRulesMap(ingFirewallConfig ingressnodefwiov1alpha1.IngressNodeFirewallRules,
-	isDelete bool, ifId uint32) error {
+// IngressNodeFwAttach attaches the eBPF program to a given list of interfaces and pins them to different pinDirs.
+// For each provided interface name:
+// i)   Look up the network interface by name.
+// ii)  Attach the program to the interface.
+// iii) Pin the XDP program.
+func (infc *IngNodeFwController) IngressNodeFwAttach(ifacesName ...string) error {
 	objs := infc.objs
-	rules := BpfRulesValSt{}
+	for _, ifaceName := range ifacesName {
+		if _, ok := infc.links[ifaceName]; ok {
+			klog.Infof("Interface %s is already attached and managed, skipping", ifaceName)
+			return nil
+		}
+		// Look up the network interface by name.
+		iface, err := net.InterfaceByName(ifaceName)
```

Comment:
use the interfaces method to get ifId and use with the Attach

Reply:
done